src/Microsoft.ML.AutoML/Experiment/Experiment.cs

```diff
@@ -49,60 +56,122 @@ internal class Experiment<TRunDetail, TMetrics> where TRunDetail : RunDetail
             _datasetColumnInfo = datasetColumnInfo;
             _runner = runner;
             _logger = logger;
+            _experimentTimerExpired = false;
+        }
+
+        private void MaxExperimentTimeExpiredEvent(object state)
+        {
+            // If at least one model was run, end experiment immediately.
+            // Else, wait for first model to run before experiment is concluded.
+            _experimentTimerExpired = true;
+            if (_history.Any(r => r.RunSucceeded))
+            {
+                _logger.Warning("Allocated time for Experiment of {0} seconds has elapsed with {1} models run. Ending experiment...",
+                    _experimentSettings.MaxExperimentTimeInSeconds, _history.Count());
+                _currentModelMLContext.CancelExecution();
+            }
+        }
+
+        private void MainContextCanceledEvent(object state)
+        {
+            // If the main MLContext is canceled, cancel the ongoing model training and MLContext.
+            if ((_context.Model.GetEnvironment() as ICancelable).IsCanceled)
+            {
+                _logger.Warning("Main MLContext has been canceled. Ending experiment...");
```

Comment:
Will `_mainContextCanceledTimer` keep running after this? If so, the user will get multiple warnings until the model returns and the timer is garbage collected.

Perhaps could either cancel the timer, or add a boolean to track if it was run.

Reply:
Good point, added changes to properly stop timer and never restart.

Reply:
Thanks #Resolved